trie/database_wrap.go

```diff
@@ -0,0 +1,267 @@
+	Initialized(genesisRoot common.Hash) bool
+
+	// Size returns the current storage size of the memory cache in front of the
+	// persistent database layer.
+	Size() common.StorageSize
+
+	// Scheme returns the identifier of used storage scheme.
+	Scheme() string
+
+	// Close closes the trie database backend and releases all held resources.
+	Close() error
+}
+
+// Database is the wrapper of the underlying backend which is shared by different
+// types of node backend implementations as an entrypoint. It's responsible for all
+// interactions relevant with trie nodes and the node preimages.
+type Database struct {
+	config    *Config          // Configuration for trie database.
+	diskdb    ethdb.Database   // Persistent database to store the snapshot
+	cleans    *fastcache.Cache // Megabytes permitted using for read caches
+	preimages *preimageStore   // The store for caching preimages
+	backend   backend          // The backend for managing trie nodes
+}
+
+// prepare initializes the database with provided configs, but the
+// database backend is still left as nil.
+func prepare(diskdb ethdb.Database, config *Config) *Database {
+	var cleans *fastcache.Cache
+	if config != nil && config.Cache > 0 {
+		if config.Journal == "" {
+			cleans = fastcache.New(config.Cache * 1024 * 1024)
+		} else {
+			cleans = fastcache.LoadFromFileOrNew(config.Journal, config.Cache*1024*1024)
+		}
+	}
+	var preimages *preimageStore
+	if config != nil && config.Preimages {
+		preimages = newPreimageStore(diskdb)
+	}
+	return &Database{
+		config:    config,
+		diskdb:    diskdb,
+		cleans:    cleans,
+		preimages: preimages,
+	}
+}
+
+// NewDatabase initializes the trie database with default settings, namely
+// the legacy hash-based scheme is used by default.
+func NewDatabase(diskdb ethdb.Database) *Database {
+	return NewDatabaseWithConfig(diskdb, nil)
+}
+
+// NewDatabaseWithConfig initializes the trie database with provided configs.
+// The path-based scheme is not activated yet, always initialized with legacy
+// hash-based scheme by default.
+func NewDatabaseWithConfig(diskdb ethdb.Database, config *Config) *Database {
+	db := prepare(diskdb, config)
+	db.backend = hashdb.New(diskdb, db.cleans, mptResolver{})
+	return db
+}
+
+// GetReader returns a reader for accessing all trie nodes with provided
+// state root. Nil is returned in case the state is not available.
+func (db *Database) GetReader(blockRoot common.Hash) Reader {
+	return db.backend.(*hashdb.Database).GetReader(blockRoot)
+}
+
+// Update performs a state transition by committing dirty nodes contained
+// in the given set in order to update state from the specified parent to
+// the specified root. The held pre-images accumulated up to this point
+// will be flushed in case the size exceeds the threshold.
+func (db *Database) Update(root common.Hash, parent common.Hash, nodes *trienode.MergedNodeSet) error {
+	if db.preimages != nil {
+		db.preimages.commit(false)
+	}
```

Comment:
Why do we have a preimage flush here too and in commit too? Aren't we expected to do an Update first, then a Commit? Do we expect to have multiple Updates and only a few Commits? Perhaps this is for the in-memory pruning where we just keep updating without committing?

Reply:
> Do we expect to have multiple Updates and only a few Commits?

Yes. `Update` has different semantics as `Commit`. 

`Update` means we add a new layer on top and these nodes will be maintained by trie database. Usually the in-memory nodes will be flushed after a few minutes once it's mature enough, this function is called on every block. 

`Commit` means we want to forcibly flush out the state(and it's ancestor) specified by this root. It's not called for every block.

Reply:
Thx